Dockerfile

```diff
@@ -1,10 +1,46 @@
-FROM rust:1.72.1 as build
+# syntax=docker/dockerfile:1
+FROM rust:1.72.1-slim-bullseye AS build
+ARG APP_NAME=aqi-alert-bot
+WORKDIR /app
 
-COPY src/ /build/src
-COPY Cargo.toml /build/Cargo.toml
-COPY Cargo.lock /build/Cargo.lock
+SHELL ["/bin/bash", "-eo", "pipefail", "-c"]
 
-WORKDIR /build
-RUN cargo build --release
+RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
 
-CMD ["/build/target/release/aqi-alert-bot"]
\ No newline at end of file
+# Required for openssl-sys
+# hadolint ignore=DL3008
+RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
+    --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    apt-get update && \
+    apt-get --no-install-recommends install -y pkg-config libssl-dev
+
+RUN --mount=type=bind,source=src,target=src \
+    --mount=type=bind,source=Cargo.toml,target=Cargo.toml \
+    --mount=type=bind,source=Cargo.lock,target=Cargo.lock \
+    --mount=type=cache,target=/app/target/ \
+    --mount=type=cache,target=/usr/local/cargo/registry/ \
+    cargo build --locked --release && \
+    cp ./target/release/$APP_NAME /bin/server
```

Comment:
this could be de-variablized maybe

Reply:
Yeah might be worth if we had a more robust CI/CD pipeline and like repo initialization scripts but probably not worth for this particular repo

Reply:
Actually this is probably just a good pattern to have in general, it's only used once in the file but it could be used more later. Just gonna leave it for future convenience's sake.